tools/SetupFlow/DevHome.SetupFlow/ViewModels/Environments/EnvironmentCreationOptionsViewModel.cs

```diff
@@ -0,0 +1,269 @@
+
+        // Selected compute system provider  may havechanged so we need to update the adaptive card in the UI
+        // with new a adaptive card from the new provider.
+        CurProviderDetails = UpcomingProviderDetails;
+
+        IsAdaptiveCardSessionLoaded = false;
+
+        // Its possible that an extension could take a long time to load the adaptive card session.
+        // So we run this on a background thread to prevent the UI from freezing.
+        _ = Task.Run(() =>
+        {
+            var developerIdWrapper = CurProviderDetails.DeveloperIds.First();
+            var result = CurProviderDetails.ComputeSystemProvider.CreateAdaptiveCardSessionForDeveloperId(developerIdWrapper.DeveloperId, ComputeSystemAdaptiveCardKind.CreateComputeSystem);
+            UpdateExtensionAdaptiveCard(result);
+        });
+    }
+
+    /// <summary>
+    /// Gets and configures the adaptive card that will be displayed on the configure environment page.
+    /// </summary>
+    public void UpdateExtensionAdaptiveCard(ComputeSystemAdaptiveCardResult adaptiveCardSessionResult)
+    {
+        _dispatcher.TryEnqueue(() =>
+        {
+            try
+            {
+                // Reset error state and remove event handler from previous session.
+                ResetAdaptiveCardConfiguration();
+
+                if (adaptiveCardSessionResult.Result.Status == ProviderOperationStatus.Failure)
+                {
+                    _log.Error($"{adaptiveCardSessionResult.Result.DisplayMessage} - {adaptiveCardSessionResult.Result.DiagnosticText}");
+                    throw new AdaptiveCardNotRetrievedException(adaptiveCardSessionResult.Result.DisplayMessage);
+                }
+
+                ExtensionAdaptiveCardSession = new ExtensionAdaptiveCardSession(adaptiveCardSessionResult.ComputeSystemCardSession);
+                ExtensionAdaptiveCardSession.Stopped += OnAdaptiveCardSessionStopped;
+                ExtensionAdaptiveCard = new ExtensionAdaptiveCard(ElementRegistration, ActionRegistration);
+                ExtensionAdaptiveCard.UiUpdate += OnAdaptiveCardUpdated;
+
+                ExtensionAdaptiveCardSession.Initialize(ExtensionAdaptiveCard);
+            }
+            catch (Exception ex)
+            {
+                _log.Error($"Failed to get creation options adaptive card from provider {CurProviderDetails.ComputeSystemProvider.Id}.", ex);
+                SessionErrorMessage = ex.Message;
+            }
+        });
+    }
+
+    public void OnAdaptiveCardUpdated(object sender, AdaptiveCard adaptiveCard)
+    {
+        _dispatcher.TryEnqueue(async () =>
+        {
+            RenderedAdaptiveCard = GetAdaptiveCardRenderer().RenderAdaptiveCard(adaptiveCard);
+            RenderedAdaptiveCard.Action += OnRenderedAdaptiveCardAction;
+
+            // Add a small delay here for the case where the setup flow is switching between the configure environment page and the review page.
+            await Task.Delay(250);
```

Comment:
Can you elaborate on this delay. Constant delays could potentially depend on each machine and may introduce inconsistent/unexpected behavior.

Reply:
spoke offline about this, but the issue was that when this method is called Dev Home can be in the middle of a page move e.g configure environment -> review page. During that move the extension is still doing work. So we end up in a situation where the page changes e.g the review page loads, requests an adaptive card and is given the previous one that was on the Configure environments page. I'll work on removing it, I was hoping to do it without adding more events but thats probably the only way, but I'll check other ways

Reply:
I removed it, looks like some other changes I made after adding the delay, removed the need for it. Thanks